src/samplerate.cpp

```diff
@@ -0,0 +1,621 @@
+                const py::object &exc_tb) {
+    _destroy();
+  }
+};
+
+namespace {
+
+long the_callback_func(void *cb_data, float **data) {
+  CallbackResampler *cb = static_cast<CallbackResampler *>(cb_data);
+  int cb_channels = cb->get_channels();
+
+  // get the data as a numpy array
+  auto input = cb->callback();
+
+  // accessors for the arrays
+  py::buffer_info inbuf = input.request();
+
+  // end of stream is signaled by a None, which is cast to a ndarray with ndim
+  // == 0
+  if (inbuf.ndim == 0) return 0;
+
+  // set the number of channels
+  int channels = 1;
+  if (inbuf.ndim == 2)
+    channels = inbuf.shape[1];
+  else if (inbuf.ndim > 2)
+    throw std::domain_error("Input array should have at most 2 dimensions");
+
+  if (channels != cb_channels || channels == 0)
+    throw std::domain_error("Invalid number of channels in input data.");
+
+  *data = static_cast<float *>(inbuf.ptr);
+
+  return (long)inbuf.shape[0];
+}
+
+}  // namespace
+
+py::array_t<float, py::array::c_style> resample(
+    const py::array_t<float, py::array::c_style | py::array::forcecast> &input,
+    double sr_ratio, const py::object &converter_type, bool verbose) {
+  // input array has shape (n_samples, n_channels)
+  int converter_type_int = get_converter_type(converter_type);
+
+  // accessors for the arrays
+  py::buffer_info inbuf = input.request();
+
+  // set the number of channels
+  int channels = 1;
+  if (inbuf.ndim == 2)
+    channels = inbuf.shape[1];
+  else if (inbuf.ndim > 2)
+    throw std::domain_error("Input array should have at most 2 dimensions");
+
+  if (channels == 0)
+    throw std::domain_error("Invalid number of channels (0) in input data.");
+
+  const auto new_size =
+      static_cast<size_t>(std::ceil(inbuf.shape[0] * sr_ratio));
+
+  // allocate output array
+  std::vector<size_t> out_shape{new_size};
+  if (inbuf.ndim == 2) out_shape.push_back(static_cast<size_t>(channels));
+  auto output = py::array_t<float, py::array::c_style>(out_shape);
+  py::buffer_info outbuf = output.request();
+
+  // libsamplerate struct
+  SRC_DATA src_data = {
+      static_cast<float *>(inbuf.ptr),   // data_in
+      static_cast<float *>(outbuf.ptr),  // data_out
+      inbuf.shape[0],                    // input_frames
+      long(new_size),                    // output_frames
+      0,        // input_frames_used, filled by libsamplerate
+      0,        // output_frames_gen, filled by libsamplerate
+      0,        // end_of_input, not used by src_simple ?
+      sr_ratio  // src_ratio, sampling rate conversion ratio
+  };
+
+  int ret_code = src_simple(&src_data, converter_type_int, channels);
+
+  error_handler(ret_code);
+
+  // create a shorter view of the array
+  if ((size_t)src_data.output_frames_gen < new_size) {
+    out_shape[0] = src_data.output_frames_gen;
+    output = py::array_t<float, py::array::c_style>(
+        out_shape, static_cast<float *>(outbuf.ptr));
```

Comment:
I was unsure about the lifetimes of the data at `outbuf.ptr` here and elsewhere in the code. However, it turned out this creates not a view of the data but a copy instead. This is safe, but it would be a good future optimization to avoid that copy.

NB: Running the tests with address sanitizers (`-fsanitize-address`) passed without issues 🎉

Reply:
I think that we can replace this by
```c++
output.resize(out_shape);
```
As I understand it, it should avoid the copy.
Also, it is still the same object, so there should not be any data ownership problem.
Here is the definition of the `resize` function.
```c++
void resize(ShapeContainer new_shape, bool refcheck = true) {
        detail::npy_api::PyArray_Dims d
            = {// Use reinterpret_cast for PyPy on Windows (remove if fixed, checked on 7.3.1)
               reinterpret_cast<Py_intptr_t *>(new_shape->data()),
               int(new_shape->size())};
        // try to resize, set ordering param to -1 cause it's not used anyway
        auto new_array = reinterpret_steal<object>(
            detail::npy_api::get().PyArray_Resize_(m_ptr, &d, int(refcheck), -1));
        if (!new_array) {
            throw error_already_set();
        }
        if (isinstance<array>(new_array)) {
            *this = std::move(new_array);
        }
    }
```
I have also created a test to trigger this case (only in `resample`) and it looks like it works.
I have checked by comparing the value of the pointer to the buffer before and after resize (looks like they are the same).
Is there a more principled way of doing that ?